### PR TITLE
Guest-dedicated Crypto Adapters passthru (jsc#SLE-4578)

### PR DESCRIPTION
**Pull request by tbazant**
### PR creator: Description

This update describes how to use host's crypto adapters inside guest VM.
Rendered version is here
[virsh-kvm-zseries-crypto_color_en.pdf](https://github.com/SUSE/doc-sle/files/6614734/virsh-kvm-zseries-crypto_color_en.pdf)

### PR creator: Are there any relevant issues/feature requests?

* jsc#SLE-4578
* jsc#DOCTEAM-41

### PR creator: Which product versions do the changes apply to?

When opening a PR, check all versions of the documentation that your PR applies to.

- SLE 15/openSUSE Leap 15.x
  - [x] SLE 15 SP4/openSUSE Leap 15.4 *(if applicable to 15 SP4 _only_ -- do not merge yet!)*
  - [x] SLE 15 SP3/openSUSE Leap 15.3 *(current `main`, no backport necessary)*
  - [x] SLE 15 SP2/openSUSE Leap 15.2
  - [x] SLE 15 SP1
  - [ ] SLE 15 GA
- SLE 12
  - [ ] SLE 12 SP5
  - [ ] SLE 12 SP4
  - [ ] SLE 12 SP3

### PR reviewer only: Have all backports been applied?

The doc team member merging your PR will take care of backporting to older documents.
When opening a PR, do *not* set the following check box.

- [ ] all necessary backports are done

### CODE REVIEW

File: xml/libvirt_configuration_virsh.xml [hunk 1]
```diff
@@ -1708,4 +1708,225 @@ domid=3: nr_frames=3, max_nr_frames=40
    </step>
   </procedure>
  </sect1>
+ <sect1 xml:id="virsh-kvm-zseries-crypto">
+  <title>Crypto adapters pass-through to &kvm; guests on &zseries;</title>
```
Comment (lproven) [inline xml/libvirt_configuration_virsh.xml:1712]:
```suggestion
  <title>Crypto adapter pass-through to &kvm; guests on &zseries;</title>
```

File: xml/libvirt_configuration_virsh.xml [hunk 2]
```diff
@@ -1708,4 +1708,225 @@ domid=3: nr_frames=3, max_nr_frames=40
...
+
+  <sect2 xml:id="virsh-kvm-zseries-crypto-intro">
+   <title>Introduction</title>
+   <para>
+    &zseries; machines include cryptographic hardware with useful functions,
+    for example, random number generation, digital signature generation, or
```
Comment (lproven) [inline xml/libvirt_configuration_virsh.xml:1718]:
```suggestion
    &zseries; machines include cryptographic hardware with useful functions
    such as random number generation, digital signature generation, or
```

File: xml/libvirt_configuration_virsh.xml [hunk 3]
```diff
@@ -1708,4 +1708,225 @@ domid=3: nr_frames=3, max_nr_frames=40
...
+    encryption. &kvm; allows dedicating these crypto adapters to guests as
+    pass-through devices while the hypervisor cannot observe the communication
```
Comment (lproven) [inline xml/libvirt_configuration_virsh.xml:1720]:
```suggestion
    pass-through devices, in which the hypervisor cannot observe communications
```

File: xml/libvirt_configuration_virsh.xml [hunk 4]
```diff
@@ -1708,4 +1708,225 @@ domid=3: nr_frames=3, max_nr_frames=40
...
+<screen>
+&prompt.user;cat /sys/devices/vfio_ap/matrix/${uuid}/matrix
+00.0016
+</screen>
+    </step>
+    <step>
+     <para>
+      Create a new VM (refer to <xref linkend="ha-kvm-inst"/>) and wait until
+      it is initialized. Alternatively, use an existing VM. In both cases, make
+      sure it is shut down.
```
Comment (lproven) [inline xml/libvirt_configuration_virsh.xml:1863]:
```suggestion
      Either create a new VM (refer to <xref linkend="ha-kvm-inst"/>) and
      wait until it is initialized, or use an existing VM. In both cases, make
      sure the VM is shut down.
```

File: xml/libvirt_configuration_virsh.xml [hunk 5]
```diff
@@ -1708,4 +1708,225 @@ domid=3: nr_frames=3, max_nr_frames=40
...
+    </step>
+   </procedure>
+  </sect2>
+
+  <sect2 xml:id="virsh-kvm-zseries-crypto-moreinfo">
+   <title>Further reading</title>
+   <itemizedlist>
+    <listitem>
+     <para>
+      Installation of virtualization components is detailed in
```
Comment (lproven) [inline xml/libvirt_configuration_virsh.xml:1907]:
```suggestion
      The installation of virtualization components is detailed in
```

File: xml/libvirt_configuration_virsh.xml [hunk 6]
```diff
@@ -1708,4 +1708,225 @@ domid=3: nr_frames=3, max_nr_frames=40
...
+      <xref linkend="cha-vt-installation"/>.
+     </para>
+    </listitem>
+    <listitem>
+     <para>
+      <literal>vfio_ap</literal> architecture is detailed in
```
Comment (lproven) [inline xml/libvirt_configuration_virsh.xml:1913]:
```suggestion
      The <literal>vfio_ap</literal> architecture is detailed in
```

File: xml/libvirt_configuration_virsh.xml [hunk 7]
```diff
@@ -1708,4 +1708,225 @@ domid=3: nr_frames=3, max_nr_frames=40
...
+      <link xlink:href="https://www.kernel.org/doc/Documentation/s390/vfio-ap.txt"/>.
+     </para>
+    </listitem>
+    <listitem>
+     <para>
+      General outline together with a detailed procedure is described in
```
Comment (lproven) [inline xml/libvirt_configuration_virsh.xml:1919]:
```suggestion
      A general outline together with a detailed procedure is described in
```

File: xml/libvirt_configuration_virsh.xml [hunk 8]
```diff
@@ -1708,4 +1708,225 @@ domid=3: nr_frames=3, max_nr_frames=40
...
+      <link xlink:href="https://bugs.launchpad.net/ubuntu/+source/linux/+bug/1787405"/>.
+     </para>
+    </listitem>
+    <listitem>
+     <para>
+      The architecture of VFIO Mediated devices (MDEV) is detailed in
```
Comment (lproven) [inline xml/libvirt_configuration_virsh.xml:1925]:
```suggestion
      The architecture of VFIO Mediated devices (MDEVs) is detailed in
```

**Review by lproven — CHANGES_REQUESTED** [review]

**Review by lproven — APPROVED** [review]

**Review by lproven — APPROVED** [review]

File: xml/libvirt_configuration_virsh.xml [hunk 9]
```diff
@@ -1708,4 +1708,225 @@ domid=3: nr_frames=3, max_nr_frames=40
...
+ <sect1 xml:id="virsh-kvm-zseries-crypto">
+  <title>Crypto adapter pass-through to &kvm; guests on &zseries;</title>
+
+  <sect2 xml:id="virsh-kvm-zseries-crypto-intro">
+   <title>Introduction</title>
+   <para>
+    &zseries; machines include cryptographic hardware with useful functions
+    such as random number generation, digital signature generation, or
+    encryption. &kvm; allows dedicating these crypto adapters to guests as
+    pass-through devices, in which the hypervisor cannot observe communications
```
Comment (lproven) [inline xml/libvirt_configuration_virsh.xml:1720]:
```suggestion
    pass-through devices. The means that the hypervisor cannot observe communications
```

**Review by lproven — CHANGES_REQUESTED** [review]

File: xml/libvirt_configuration_virsh.xml [hunk 10]
```diff
@@ -1708,4 +1708,225 @@ domid=3: nr_frames=3, max_nr_frames=40
...
+     </para>
+<screen>
+&prompt.sudo;echo +0x0 > /sys/devices/vfio_ap/matrix/${uuid}/assign_adapter
+&prompt.user;echo +0x16 | sudo tee /sys/devices/vfio_ap/matrix/${uuid}/assign_domain
+&prompt.user;echo +0x16 | sudo tee /sys/devices/vfio_ap/matrix/${uuid}/assign_control_domain
+</screen>
+    </step>
+    <step>
+     <para>
+      verify the matrix that you have configured:
```
Comment (tbazant) [inline xml/libvirt_configuration_virsh.xml:1852]:
```suggestion
      Verify the matrix that you have configured:
```

**Review by lproven — APPROVED** [review]

**Comment by tbazant** [comment]
> Just one small place: Could you add driver vfio_ap and vfio_mdev to Requirements15.18.3. Move 4 to 1 in 15.18.4 by checking driver with "lsmod | grep vfio_", make sure if vfio_ap and vfio_mdev are loaded, otherwise load them with modprobe.
> 
> I also saw there are a lot of operations with "bash code" here, just wondering if you tested them in our SLE environment? sometimes the output may be different which will confuse our readers.

@liayan i have included your suggestion. and no, i havent tested it

**Review by lproven — APPROVED** [review]

**Pull request merged**